### PR TITLE
docker_client moved into dockerapi pkg

### PR DESCRIPTION
**Pull request by yunhee-l**
### Summary
docker_client moved into agent/dockerapi pkg along with related files, enginError replaced with NamedError

### Implementation details
1. docker_client.go moved from engine/ to dockerclient/. This required a lot of constants and structs to be exported, as engine/ needed to access.
2. other related files (including corresponding _test.go files) moved from engine/ to dockerclient/:
docker_events_buffer.go
repotag.go
types.go 
3. engine/errors split into docker-related error and agent-side error, and moved into dockerclient/
4. engine.engineError is a duplicate of api.NamedError. api.NamedError is moved to api/errors package 
5. MetadataFromContainer function in docker_client.go refactored to lower cyclomatic complexity and pass the gocyclo static check

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [ ] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/dockerclient/dockerapi/docker_client.go [hunk 1]
```diff
@@ -67,11 +67,11 @@ const (
 	// TODO: Benchmark and re-evaluate this value
 	LoadImageTimeout        = 10 * time.Minute
 	pullImageTimeout        = 2 * time.Hour
-	createContainerTimeout  = 4 * time.Minute
-	stopContainerTimeout    = 30 * time.Second
-	removeContainerTimeout  = 5 * time.Minute
-	inspectContainerTimeout = 30 * time.Second
-	removeImageTimeout      = 3 * time.Minute
+	CreateContainerTimeout  = 4 * time.Minute
```
Comment (aaithal) [inline agent/dockerclient/dockerapi/docker_client.go]: lint: Can you add documentation for all of the globally visible const's? See https://github.com/golang/go/wiki/CodeReviewComments#doc-comments and https://golang.org/doc/effective_go.html#commentary for more details.

File: agent/dockerclient/dockerapi/errors.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+import (
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient"
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+const (
+	DockerTimeoutErrorName           = "DockerTimeoutError"
```
Comment (aaithal) [inline agent/dockerclient/dockerapi/errors.go]: here too, missing documentation for consts

File: agent/dockerclient/dockerapi/errors.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+	CannotInspectContainerErrorName  = "CannotInspectContainerError"
+	CannotDescribeContainerErrorName = "CannotDescribeContainerError"
+)
+
+// DockerTimeoutError is an error type for describing timeouts
+type DockerTimeoutError struct {
+	Duration   time.Duration
```
Comment (aaithal) [inline agent/dockerclient/dockerapi/errors.go]: I know you're just moving these between files. But, this is as good a time as any to add comments for these fields.

File: agent/utils/errors.go [hunk 4]
```diff
@@ -18,6 +18,12 @@ import (
 	"strings"
 )
 
+// NamedError defines an interface that wraps error and add additional 'ErrorName' method
+type NamedError interface {
```
Comment (aaithal) [inline agent/utils/errors.go]: What do you think about moving this to "api/errors" package? The generic 'utils' package doesn't seem like a good home for this.

Comment (yunhee-l) [reply agent/utils/errors.go]: I've moved this file into api/errors/error_types.go.

**Pull request merged**